### ISSUE TITLE
Improve screen reader accessibility

### ISSUE THREAD

**Issue by Flameborn** [izderadicka/audioserve-web#10]
Hello,

This new web client works beautifully with screen readers (such as [NVDA](https://nvda-project.org) on Windows) when using the keyboard only.

There are a few minor things, however, that could be improved if possible, for the player component:

1. There are *div* elements that can be clicked on, such as the play button. These need:
  * A button ARIA role
  * An ARIA label (in the case of the play/pause button either Play or Pause, depending on active playback)
  * A tabIndex, so that they can be tabbed to and receive focus via the keyboard. A -1 value should suffice
  * In the case of the expandable button, an ARIA expanded state, either true or false based on the expand state, which changes as the button is toggled
2. Sliders, e.g. volume, speed and position need an ARIA label
3. For the folder browser, some indication that the headings are clickable would be nice to have. This could be accomplished by including the heading title within a span, having the link role, for example.

Thank you very much for any possible improvement.

**Comment by Ghorthalon** [comment]
I've been able to fix these issues locally. I'll open a pull request soon.

**Comment by Flameborn** [comment]
Just to make sure that there's a link to the related PR, https://github.com/izderadicka/audioserve-web/pull/11 fixes what I could find. The hotkey disabling logic, as mentioned in the PR, as well as possibly adding arrow key navigation to the folder list needs further discussion as it likely requires input from @izderadicka.

**Comment by izderadicka** [comment]
@Flameborn  @Ghorthalon

Hi guys,  first thanks a lot for rising the issue and also for PR.  I definitely will look at it and might require your further help. 
I have to say I'm quite pathetic UI designer/programmer and my knowledge of Accessibility design principles / practices  is very basic, but I do understand it is important.  
I'll be  off  most time next 3 weeks, so I guess I'll get to it sometime around mid July.  If you'll have any further inputs please add it to this issue or as further commits to PR #11 .

Concerning keyboard navigation I recently (over weekend) added basic navigation - space for play/pause and arrows for short jumps within track/chapter.

**Comment by izderadicka** [comment]
@Ghorthalon 

As I have written in #11 now there is new branch accessibility with your changes.   
They do have visual effect,  `role=button`  add blue rectangle around element,  so bit more styling will be needed..  I'm using picocss framework.

As you have written it's better to have button element then span element with button role I think we should replace spans with buttons, but this will need special style - something like span-button, which will not apply any default button styling. , 

What do you think?

**Comment by izderadicka** [comment]
I've done some  changes on branch accessibility - see commit a25cfa0475d268c4944794abc08947c59b7e0948 - mainly to keep visual layout the same after adding roles to some elements. 

Right now I did not replace span-like buttons with real buttons,   mainly because it'll infer now with global key - bindings, where space is used for play/pause.  But during work on this commit I realized that this new  global keyboard commands may interfere with accessibility.  

I'd definitely like to hear your feedback on branch accessibility,  if it works for you and what else need to be done.

**Comment by Flameborn** [comment]
Thanks for this.

From a screen reader user's perspective, visual styling rarely affects the various elements, since all a screen reader needs to know is what type of control an element is. A button, for example, will be a button regardless of having a border.

The aim of @Ghorthalon's PR is to add roles that screen readers understand, as well as make keyboard navigation easier, while, hopefully, not modifying how a specific page appears.

In addition, there were some elements that had an icon only, these received ARIA labels, so that screen reader users who can't see the images may know what these controls do.

In addition to what was posted in the PR, I've made some more changes, please let me know if I should open a separate pull request instead. Unfortunately I can't modify the PR and these changes were not merged, but @Ghorthalon's PR can be patched.

The patch does the following:

* Adds a button role and label for the sort button
* Makes the download link visible by adding an ARIA label
* Labels the close element, e.g. for the recently listened popup
* Labels the *Shared Secret* and the *Playback Group* fields on the login page
* Changes the label for the *more controls* button, because later on, other extra controls might be added

[fixes.patch.zip](https://github.com/izderadicka/audioserve-web/files/9034572/fixes.patch.zip)

If everything that @Ghorthalon added, as well as this patch is included via the accessibility branch, every element becomes visible and can be used via a screen reader.

Regarding the global keyboard shortcuts, traditionally, if a button receives focus, a user should be able to press it via the space key, but I think that if every button receives a global hotkey, this is not as important.

**Comment by izderadicka** [comment]
@Flameborn 
All changes from  @Ghorthalon  PR  #11 are included  in accessibility branch. If you can please add your changes as PR against that branch. There are also couple more additional changes I've added in hope to improve accesibility.

Concerning. styling  my problem  was that picocss is styling elements with role button same as  actual button element, so adding role changes visual appearance significantly by adding background, borders etc. to element. I mitigated this by an additional class and rules.

I was just thinking what about an interactive session via Zoom or similar. We can go through application and  how you use it and discuss how to improve?

**Comment by pitermach** [comment]
Thank you for all your work on this and considering accessibility. I've really been enjoying using audio serve and find it extremely helpful!

Regarding global hotkeys, space for play/pause is probably fine if you made it so that it works as long as the focus isn't on one of the other buttons. If you wanted to free up the arrow keys, you could adopt shortcuts similar to the ones on the YouTube and soundCloud player. Something like the following:

- J K L = rewind, play/pause, fast forward
- Shift+Down/up -- volume down/up.
- Shift+. and Shift+, - adjust speed
- 1...0 - go to 10-100% of the file
- Shift+N/P - next/previous file

That would let you free up the arrow keys if you wanted them to navigate through the list of files/folders like @Ghorthalon suggested in his PR.

Lastly two small suggestions I have for the player, I think it should be put in its own aria region so you can reach it faster (screen readers on both PC and mobile have a way to quickly jump between regions). Also, for the position slider, I think you should give it a [aria-valuetext](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-valuetext) giving the current and total time in the current file. Right now, screen readers always read the position slider in seconds and you need to look at the surrounding text to see the time in a more friendly format. In contrast, the position sliders in most media players usually report the time to screen readers on the slider itself - IE "02:35 of 22:20)

**Comment by izderadicka** [comment]
@pitermach 
Thanks
considering global hot keys I'm still doubting - if space can be consumed by focused button it will be colliding with global binding cause most of time some button will be focused. Other keys are fine just was not sure what should be "standard". So for now just mapped media keys plus space and left, right arrows.

Player as region - absolutely! what would be right aria attribute?  I was thinking about making info, folders , files and player regions so you can  quickly navigate between them. Or player and browser on same level?

aria-valuetext - cool - I'll add it.

**Comment by Flameborn** [comment]
I've just pushed a further improvements commit via https://github.com/izderadicka/audioserve-web/pull/12

Regarding regions, yes, I think having one for each logical part of the application is a good idea for quicker navigation, especially on phones.

Although I'm not using Audioserve every day yet, mainly due to the speed algorithm on iOS, I think now the only thing we can possibly improve is the folder/file browser, to make it behave like a native app via the keyboard, similarly to how e.g. the Dropbox or Google Drive web apps work.

For this, the arrow keys could navigate between items (folders and files), enter open the current one, and possibly backspace go up a level.

Though I'd love to see this, I have to mention that this is more of a convenience feature, since the client is usable once the accessibility branch is merged.

Thanks for offering a Zoom session to discuss further accessibility improvements. Currently, other than what was mentioned, I'm not sure what else could be improved, but I'd be more than happy to discuss how we use the client if this helps. I really appreciate your openness regarding this.

**Comment by izderadicka** [comment]
@Flameborn  - concerning iOS and Apple stuff - I recently added some small improvements for kind better compatibility, main problem is opus codec, which is defaulted for transcoding, but not supported (with some exceptions on MacOS)  - check [this new section in audioserve readme](https://github.com/izderadicka/audioserve#alternative-transcodings-and-transcoding-configuration-for-apple-users).  Also in new client you can then enforce transcoding for mime types not supported - audio/ogg etc.

What do you mean by "speed algorithm on iOS"? Is there anything we can do here?

**Comment by Flameborn** [comment]
Ah yes, at the moment I just switch the transcoding to *none* for the clients when I log in, but I should probably set up an m4a transcoding profile.

Safari's built-in speed algorithm is the one used by Quicktime, which is unfortunately very poor. Other browsers, such as Chrome and Firefox use Apple's more modern one, which sounds a lot natural. Under iOS, since every browser has to use Webkit, the speed algorithm is the same as in Safari.

I looked into something like Sonic, but it does not have a web version, so I'll probably have to use something like Emscripten, but that's for the future.

**Comment by izderadicka** [comment]
@Flameborn - for AAC transcoding audioserve now supports only ADTS container -  "classic" MPEG4 container does not work for streamed data (which is what we are doing , file is streamed to ffmpeg via stdin pipe and then directly to http response body).

**Comment by izderadicka** [comment]
I've merged accessibility branch to main - it contains PR from @Ghorthalon  @Flameborn  (thanks again) and some of my changes, including changes proposed by @pitermach - aria-valuetext and regions . I hope my changes are useful, if you find some issues around them let me know - for instance for some aria-labels I made them more descriptive, but maybe for screen reader it's better if they are short, I added some other aria-labels, when I felt it might be useful etc.. But often less is more - so do not hesitate to provide feedback, if changes are rather disruptive then useful.

You can test here https://audioserve-new.zderadicka.eu/ 
(@Flameborn - server is configured for aac transcoding for iPhone/IPad, as I noted in previous comments).

What is left is:
- [Extend global key biding] (https://github.com/izderadicka/audioserve-web/issues/10#issuecomment-1173077170)
- Keyboard navigation in Files/Folders lists

It will require bit more thoughts on my side, so it will not be available immediately. And I also want to spend some time on server, as Rust programming is more fun :-). 

Please share any other ideas or wishes for accessibility improvements - I'll keep this issue open.

**Comment by Flameborn** [comment]
This is perfect, thank you. It also turns out I’ve missed the locate playing file and navigate to… elements, good catch!

Transcoding works in Safari now as well, tested under MacOS.

There’s a small typo: "Posion of currently playing file in folder” in src/components/Player.svelte.

Oh yes, I can certainly understand why coding in Rust is more fun, I don’t code much in JS unless I have to ☺️

Thanks again for all your work and for possibly adding the global hotkeys and arrow key navigation in the future, whenever you can.

**Comment by pitermach** [comment]
A big thanks from me as well. I wanted to get back to you regarding the regions, but you beat me to it and I think you nailed them perfectly! :)

**Comment by Ghorthalon** [comment]
Hey, sorry for being quiet on this issue, however it seems like everything was able to proceed without me just fine 😁 

Thanks so much for taking all this into consideration. The new demo works fine, and I really love using this now. Everything else is a bonus and not absolutely required. If you still want to talk to someone about how we actually use the player, I'd be happy to either join a meeting or just record a little demo.

Thanks for all your work!
